### PR TITLE
BookA-Pset_SpatialGeoObsBorehole

### PR DESCRIPTION
**Pull request by larswik**
# Pset: Pset_SpatialGeoObsBorehole

## Description: 

Properties for geoscience observations in boreholes

| Property | Datatype | Description |
|----------|----------|-------------|
|BoreholeName|IfcLabel|Identifier of a borehole like No. and name  - Not part of the Pset/Qto|
|BoreholeType|PEnum_BoreholeType|Type of drilling methods used to perform the investigation|
|BoreholeElevation|IfcLengthMeasure|If present, the property reports the elevation data, in metres, for the borehole (i.e., wellbore) start point. This is a compromise approach to allow for delivery of legacy 2D data without elevation data, and for software that cannot process a 3D GM_Point.|
|BoreholeInclination|IfcPlaneAngleMeasure|"The angle formed by a vertical line and the vertical component in the borehole direction is entered. "|
|BoreholeAzimuth|IfcPlaneAngleMeasure|"If the drilling is inclined, the horizontal component in the borehole direction is entered based on the 360�orientation method moving right from true north. "|
|BoreholePurpose|IfcLabel|The purpose property reports the purpose for which the borehole was drilled.|
|BoreholeState|PEnum_BoreholeState|The property status reports the current status of the borehole.|
|Operator|IfcLabel|The property operator reports the organisation or agency responsible for commissioning of the borehole (as opposed to the agency which drilled the borehole).|
|Driller|IfcLabel|The property driller reports the organisation responsible for drilling the borehole (as opposed to commissioning the borehole).|
|BoreholeLengthPlanned|IfcPositiveLengthMeasure|The property boreholeLengthPlanned reports the planned length of a borehole e.g. in a site investigation program, in metres, as determined by the data provider. |
|BoreholeLengthDrilled|IfcPositiveLengthMeasure|The property boreholeLengthDrilled reports the length of a borehole as drilled and logged, in metres, as determined by the data provider. |
|DrillingMethod|IfcLabel|"Type of drilling methods used to perform the investigation, "|

## Enumeration: PEnum_BoreholeType

| Constant | Description |
|----------|-------------|
|COREDRILLING|A drilling process in which borehole is performed by cutting out cylindrical rock or soil samples in the field, using a core barrel.|
|DESTRUCTIVEDRILLING|A drilling process in which the boring is performed using destructive tools|
|TRIALPIT|An excavation made for the purpose of observing shallow subsurface conditions, performing field tests and obtaining soil samples.|
|OTHER|Another, user defined, type of borehole.|

## Enumeration: PEnum_BoreholeState

| Constant | Description |
|----------|-------------|
|Existing list (IFC 4.3)| |

### CODE REVIEW

**Comment by SergejMuhic** [comment]
Constants are without descriptions.

**Comment by larswik** [comment]
> Constants are without descriptions.

Should be there with latest commit.

File: IFC4x3/Properties/b/BoreholeAzimuth_0gN7ztT89FlOcDscg5dCUo/Documentation.md
Comment (SergejMuhic) [inline IFC4x3/Properties/b/BoreholeAzimuth_0gN7ztT89FlOcDscg5dCUo/Documentation.md:1]: I would say this is derivable from geometry. I would not add such properties.

File: IFC4x3/Properties/b/BoreholeElevation_0zNwWthFTDjuOnS8GoqDch/DocProperty.xml
Comment (SergejMuhic) [inline IFC4x3/Properties/b/BoreholeElevation_0zNwWthFTDjuOnS8GoqDch/DocProperty.xml:1]: Same as previous comment.

File: IFC4x3/Properties/b/BoreholeElevation_0zNwWthFTDjuOnS8GoqDch/Documentation.md [hunk 1]
```diff
@@ -0,0 +1 @@
+If present, the property reports the elevation data, in metres, for the borehole (i.e., wellbore) start point. This is a compromise approach to allow for delivery of legacy 2D data without elevation data, and for software that cannot process a 3D GM_Point.
```
Comment (SergejMuhic) [inline IFC4x3/Properties/b/BoreholeElevation_0zNwWthFTDjuOnS8GoqDch/Documentation.md]:
```suggestion
If present, the property reports the elevation data in meters for the borehole (i.e., wellbore) start point. This is a compromise approach to allow for delivery of legacy 2D data without elevation data, and for software that cannot process a 3D GM_Point.
```

Comment (SergejMuhic) [reply IFC4x3/Properties/b/BoreholeElevation_0zNwWthFTDjuOnS8GoqDch/Documentation.md]: IFC file do not come in 2D geometry only.

File: IFC4x3/Properties/b/BoreholeInclination_0DT5srGsL6xA1DS1X9AiCL/DocProperty.xml
Comment (SergejMuhic) [inline IFC4x3/Properties/b/BoreholeInclination_0DT5srGsL6xA1DS1X9AiCL/DocProperty.xml:1]: Geometry property again.

File: IFC4x3/Properties/b/BoreholeLengthDrilled_3Q4N7zz8fFIPA03iGeIzS1/Documentation.md [hunk 2]
```diff
@@ -0,0 +1 @@
+The length of a borehole as drilled and logged, in metres, as determined by the data provider.
```
Comment (SergejMuhic) [inline IFC4x3/Properties/b/BoreholeLengthDrilled_3Q4N7zz8fFIPA03iGeIzS1/Documentation.md]:
```suggestion
The length of a borehole as drilled and logged, in meters, as determined by the data provider. 
```

File: IFC4x3/Properties/b/BoreholeLengthPlanned_1_khbjyjf8z9JjevtGbGtr/Documentation.md [hunk 3]
```diff
@@ -0,0 +1 @@
+The planned length of a borehole e.g. in a site investigation program, in metres, as determined by the data provider.
```
Comment (SergejMuhic) [inline IFC4x3/Properties/b/BoreholeLengthPlanned_1_khbjyjf8z9JjevtGbGtr/Documentation.md]:
```suggestion
The planned length of a borehole e.g. in a site investigation program, in meters, as determined by the data provider. 
```

File: IFC4x3/Properties/b/BoreholeState_2isSw4SEv80elc7lNb5yJI/Documentation.md
Comment (SergejMuhic) [inline IFC4x3/Properties/b/BoreholeState_2isSw4SEv80elc7lNb5yJI/Documentation.md:1]: For the sake of consistency, use 'state' or 'status' in documentation and property name.

File: IFC4x3/Properties/b/BoreholeType_0Cyl632AvEDgJ$ALRGfq6u/Documentation.md
Comment (SergejMuhic) [inline IFC4x3/Properties/b/BoreholeType_0Cyl632AvEDgJ$ALRGfq6u/Documentation.md:1]: The description implies that this property should be named "DrillingMethod" and not "BoreholeType".

File: IFC4x3/Properties/d/Driller_3Pf78hscf4NOQSyScVWVgk/DocProperty.xml
Comment (SergejMuhic) [inline IFC4x3/Properties/d/Driller_3Pf78hscf4NOQSyScVWVgk/DocProperty.xml:1]: Consider renaming to "DrillingContractor" or something in that manner to be more precise.

File: IFC4x3/Properties/d/DrillingMethod_2tlZTCJoP5V9xcumCFudxO/DocProperty.xml
Comment (SergejMuhic) [inline IFC4x3/Properties/d/DrillingMethod_2tlZTCJoP5V9xcumCFudxO/DocProperty.xml:1]: Seems like a duplication to `BoreholeType`.

File: IFC4x3/Properties/o/Operator_39FqXYJQT2U9LFyRW241DI/Documentation.md [hunk 4]
```diff
@@ -0,0 +1 @@
+The property operator reports the organisation or agency responsible for commissioning of the borehole (as opposed to the agency which drilled the borehole).
```
Comment (SergejMuhic) [inline IFC4x3/Properties/o/Operator_39FqXYJQT2U9LFyRW241DI/Documentation.md]:
```suggestion
The organisation or agency responsible for commissioning the borehole (as opposed to the agency which drilled the borehole).
```

File: IFC4x3/PropertyConstants/c/COREDRILLING_3kAMQZBc19h9RiBVqacIVy.xml [hunk 5]
```diff
@@ -0,0 +1,5 @@
+<?xml version="1.0" encoding="utf-8"?>
+<DocPropertyConstant xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" id="COREDRILLING_3kAMQZBc19h9RiBVqacIVy" Name="COREDRILLING" UniqueId="ee2966a3-2e60-49ac-96ec-2dfd249927fc">
+	<Documentation>A drilling process in which borehole is performed by cutting out cylindrical rock or soil samples in the field, using a core barrel.</Documentation>
```
Comment (SergejMuhic) [inline IFC4x3/PropertyConstants/c/COREDRILLING_3kAMQZBc19h9RiBVqacIVy.xml]:
```suggestion
	<Documentation>A drilling process in which borehole is performed by cutting out cylindrical rock or soil samples in the field using a core barrel.</Documentation>
```

File: IFC4x3/PropertyConstants/o/OTHER_1e965qwjTCJP79vQ7RW1XB.xml [hunk 6]
```diff
@@ -0,0 +1,5 @@
+<?xml version="1.0" encoding="utf-8"?>
+<DocPropertyConstant xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" id="OTHER_1e965qwjTCJP79vQ7RW1XB" Name="OTHER" UniqueId="68246174-ead7-4c4d-91c9-e5a1db80184b">
+	<Documentation>Another, user defined, type of borehole.</Documentation>
```
Comment (SergejMuhic) [inline IFC4x3/PropertyConstants/o/OTHER_1e965qwjTCJP79vQ7RW1XB.xml]: Why not name it USERDEFINED then?
```suggestion
	<Documentation>User defined, type of borehole.</Documentation>
```

File: IFC4x3/PropertyConstants/r/REFILLED_14xFmF1EX5c96JpnBSWrNe.xml
Comment (SergejMuhic) [inline IFC4x3/PropertyConstants/r/REFILLED_14xFmF1EX5c96JpnBSWrNe.xml:1]: Description.

File: IFC4x3/PropertyConstants/t/TRIALPIT_3A22i0_h53XBG31iFFTzGQ.xml [hunk 7]
```diff
@@ -0,0 +1,5 @@
+<?xml version="1.0" encoding="utf-8"?>
+<DocPropertyConstant xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" id="TRIALPIT_3A22i0_h53XBG31iFFTzGQ" Name="TRIALPIT" UniqueId="ca082b00-fab1-4384-b403-06c3cf77d41a">
+	<Documentation>An excavation made for the purpose of observing shallow subsurface conditions, performing field tests and obtaining soil samples.</Documentation>
```
Comment (SergejMuhic) [inline IFC4x3/PropertyConstants/t/TRIALPIT_3A22i0_h53XBG31iFFTzGQ.xml]:
```suggestion
	<Documentation>An excavation made for the purpose of observing shallow subsurface conditions performing field tests and obtaining soil samples.</Documentation>
```

**Review by SergejMuhic — CHANGES_REQUESTED** [review]
In general, I would skip "Borehole" in borehole property names.

**Comment by SergejMuhic** [comment]
@michel-rives please look at the comments so that I can resolve and merge.

File: IFC4x3/Properties/b/BoreholeAzimuth_0gN7ztT89FlOcDscg5dCUo/Documentation.md
Comment (Michel-Rives) [reply IFC4x3/Properties/b/BoreholeAzimuth_0gN7ztT89FlOcDscg5dCUo/Documentation.md:1]: Please leave as is

**Comment by Michel-Rives** [comment]
> In general, I would skip "Borehole" in borehole property names.

Please leave the Borehole prefix

File: IFC4x3/Properties/b/BoreholeElevation_0zNwWthFTDjuOnS8GoqDch/Documentation.md (hunk 1 above)
Comment (Michel-Rives) [reply IFC4x3/Properties/b/BoreholeElevation_0zNwWthFTDjuOnS8GoqDch/Documentation.md]: The change proposed ('meters') is fine

File: IFC4x3/Properties/b/BoreholeLengthDrilled_3Q4N7zz8fFIPA03iGeIzS1/Documentation.md (hunk 2 above)
Comment (Michel-Rives) [reply IFC4x3/Properties/b/BoreholeLengthDrilled_3Q4N7zz8fFIPA03iGeIzS1/Documentation.md]: The change proposed ('meters') is fine

File: IFC4x3/Properties/b/BoreholeLengthPlanned_1_khbjyjf8z9JjevtGbGtr/Documentation.md (hunk 3 above)
Comment (Michel-Rives) [reply IFC4x3/Properties/b/BoreholeLengthPlanned_1_khbjyjf8z9JjevtGbGtr/Documentation.md]: The change proposed ('meters') is fine

File: IFC4x3/Properties/b/BoreholeState_2isSw4SEv80elc7lNb5yJI/Documentation.md
Comment (Michel-Rives) [reply IFC4x3/Properties/b/BoreholeState_2isSw4SEv80elc7lNb5yJI/Documentation.md:1]: Please change the name of the property to 'BoreholeStatus'

File: IFC4x3/Properties/b/BoreholeType_0Cyl632AvEDgJ$ALRGfq6u/Documentation.md
Comment (Michel-Rives) [reply IFC4x3/Properties/b/BoreholeType_0Cyl632AvEDgJ$ALRGfq6u/Documentation.md:1]: Please change the description to "Type of borehole as a result of the drilling methos used"

File: IFC4x3/Properties/d/Driller_3Pf78hscf4NOQSyScVWVgk/DocProperty.xml
Comment (Michel-Rives) [reply IFC4x3/Properties/d/Driller_3Pf78hscf4NOQSyScVWVgk/DocProperty.xml:1]: Please change the property name to 'DrillingContractor'
Please change the description to 'Specifies the organization responsible for the drilling (as opposed to commissioning the drilling)'

File: IFC4x3/Properties/d/DrillingMethod_2tlZTCJoP5V9xcumCFudxO/DocProperty.xml
Comment (Michel-Rives) [reply IFC4x3/Properties/d/DrillingMethod_2tlZTCJoP5V9xcumCFudxO/DocProperty.xml:1]: Please leave as is

File: IFC4x3/Properties/o/Operator_39FqXYJQT2U9LFyRW241DI/Documentation.md (hunk 4 above)
Comment (Michel-Rives) [reply IFC4x3/Properties/o/Operator_39FqXYJQT2U9LFyRW241DI/Documentation.md]: Proposed change is fine

File: IFC4x3/PropertyConstants/c/COREDRILLING_3kAMQZBc19h9RiBVqacIVy.xml (hunk 5 above)
Comment (Michel-Rives) [reply IFC4x3/PropertyConstants/c/COREDRILLING_3kAMQZBc19h9RiBVqacIVy.xml]: Proposed change is fine

File: IFC4x3/PropertyConstants/o/OTHER_1e965qwjTCJP79vQ7RW1XB.xml (hunk 6 above)
Comment (Michel-Rives) [reply IFC4x3/PropertyConstants/o/OTHER_1e965qwjTCJP79vQ7RW1XB.xml]: Please change the description to 'User defined, other type of borehole'
Please leave the constant 'OTHER' as is

File: IFC4x3/PropertyConstants/r/REFILLED_14xFmF1EX5c96JpnBSWrNe.xml
Comment (Michel-Rives) [reply IFC4x3/PropertyConstants/r/REFILLED_14xFmF1EX5c96JpnBSWrNe.xml:1]: Please remove

File: IFC4x3/PropertyConstants/t/TRIALPIT_3A22i0_h53XBG31iFFTzGQ.xml (hunk 7 above)
Comment (Michel-Rives) [reply IFC4x3/PropertyConstants/t/TRIALPIT_3A22i0_h53XBG31iFFTzGQ.xml]: Proposed change is fine

**Comment by SergejMuhic** [comment]
Final touch before merge. I am unsure on what to do with `BoreholeState` that was changed to `BoreholeStatus`. They seem to reference the same enum but have a different description and are in different Pset. The present one and "Pset_BoreholeCommon".

@larswik @Michel-Rives @JontyDunn @j-weil please advise.

**Comment by larswik** [comment]
1. We seem to be missing PEnum_BoreholeState in our documentation - Ping @SergejMuhic 
2. I would suggest removing BoreholeState/BoreholeStatus from this Pset since it is already in Pset_BoreholeCommon. Since both properties re-use the same enum-list, I don't see that there could be different definitions? - Ping @j-weil
3. Regarding the property BoreholeType (Ping @j-weil):
    a. Remove the BoreholeType property since it completely overlaps the predefined type enumeration in [IfcBoreholeTypeEnum](https://bsi-infraroom.github.io/IFC-Documentation-Tunnel/4_4_0_0/general/HTML/link/ifcboreholetypeenum.htm). Keeping the list as predefined types gives the possibility to attach/require custom Psets per type. 
    b. In IFC 4.3, IfcBorehole does not have a PredefinedType at all so an alternative would be to remove the PredefinedType attribute added by the Tunnel project (and be backwards compatible with 4.3 - IfcGeoScienceFeature doesn't add attributes) and keep the BoreholeType property. However, what does the guidelines have to say when IfcElement decendants have no PredefinedType?

**Comment by larswik** [comment]
Regarding the issue of BoreholeType property vs a corresponding PredefinedType, the tech team discussed and proposes to keep the PredefinedType and remove the BoreholeType property. Reason being PredefinedType is a standard IFC typing mechanism that allows for specifying PDT specific Pset applicability and that should be available "everywhere".
Ping @j-weil

BoreholeState property should be removed since it is part of Pset_BopreholeCommon.
Regarding PEnum_BoreholeState, see [PEnum_BoreholeStatus](https://github.com/bSI-InfraRoom/IFC-Specification/blob/cd0378df6a4fead4c92a590f7a218017e0d783c1/IFC4x3/PropertyEnumerations/b/PEnum_BoreholeStatus.xml).

**Comment by larswik** [comment]
Last update:
- Remove property BoreholeType entirely (including PEnum_BoreholeType and its constants)
- Remove property BoreholeStatus from this Pset (and delete the BoreholeStatus property - In Pset_BoreholeCommon, the property is still BoreholeState - from IFC 4.3)
image

**Pull request merged**